Clean up Kinesis doc (#14529)
apache/druid: docs/development/extensions-core/kinesis-ingestion.md

  ~ under the License.
  -->

When you enable the Kinesis indexing service, you can configure supervisors on the Overlord to manage the creation and lifetime of Kinesis indexing tasks. These indexing tasks read events using Kinesis' own shard and sequence number mechanism to guarantee exactly-once ingestion. The supervisor oversees the state of the indexing tasks to coordinate handoffs, manage failures, and ensure that scalability and replication requirements are maintained.

This topic contains configuration reference information for the Kinesis indexing service supervisor for Apache Druid.

## Setup

To use the Kinesis indexing service, you must first load the `druid-kinesis-indexing-service` core extension on both the Overlord and the Middle Manager. See [Loading extensions](../../configuration/extensions.md#loading-extensions) for more information.

> We recommend that you review the [Kinesis known issues](#kinesis-known-issues) before deploying the `druid-kinesis-indexing-service` extension to production.

Reviewer: ```suggestion
Review the [Kinesis known issues](#kinesis-known-issues) before deploying the `druid-kinesis-indexing-service` extension to production.
```

To create a supervisor, send a `POST` request to the `/druid/indexer/v1/supervisor` endpoint.
Once created, the supervisor persists in the configured metadata database. There can only be a single supervisor per datasource, and submitting a second spec for the same datasource overwrites the previous one.

When an Overlord gains leadership, either by being started or as a result of another Overlord failing, it spawns
a supervisor for each supervisor spec in the metadata database. The supervisor then discovers running Kinesis indexing
tasks and attempts to adopt them if they are compatible with the supervisor's configuration. If they are not
compatible because they have a different ingestion spec or shard allocation, the tasks are killed and the
supervisor creates a new set of tasks. In this way, the supervisors persist across Overlord restarts and failovers.

The following example shows how to submit a supervisor spec for a stream with the name `KinesisStream`.
In this example, `http://SERVICE_IP:SERVICE_PORT` is a placeholder for the server address of deployment and the service port.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->
```shell

> curl -X POST "http://SERVICE_IP:SERVICE_PORT/druid/indexer/v1/supervisor" \

Reviewer: normally we're running all HTTP API calls through the router. Retest it with this :
`curl -X POST "http://SERVICE_IP:SERVICE_PORT/druid/router/v1/supervisor"`

`

      "dataSource": "KinesisStream",
      "timestampSpec": {
        "column": "timestamp",
        "format": "iso"
      },
      "dimensionsSpec": {
        "dimensions": [
          "isRobot",
          "channel",
          "flags",
          "isUnpatrolled",
          "page",
          "diffUrl",
          {
            "type": "long",
            "name": "added"
          },
          "comment",
          {
            "type": "long",
            "name": "commentLength"
          },
          "isNew",
          "isMinor",
          {
            "type": "long",
            "name": "delta"
          },
          "isAnonymous",
          "user",
          {
            "type": "long",
            "name": "deltaBucket"
          },
          {
            "type": "long",
            "name": "deleted"
          },
          "namespace",
          "cityName",
          "countryName",
          "regionIsoCode",
          "metroCode",
          "countryIsoCode",
          "regionName"
        ]
      },
      "granularitySpec": {
        "queryGranularity": "none",
        "rollup": false,
        "segmentGranularity": "hour"
      }
    }
  }
}'
```
<!--HTTP-->
```HTTP

> POST /druid/indexer/v1/supervisor

Reviewer: same comment as line 65 wrt/ router vs indexer

The following table outlines the configuration options for `autoScalerConfig`:

|Property|Description|Required|Default|
|--------|-----------|--------|-------|
|`enableTaskAutoScaler`|Enables the auto scaler. If not specified, Druid disables the auto scaler even when `autoScalerConfig` is not null.|No|`false`|
|`taskCountMax`|Maximum number of Kinesis ingestion tasks. Must be greater than or equal to `taskCountMin`. If greater than `{numKinesisShards}`, Druid sets the maximum number of reading tasks to `{numKinesisShards}` and ignores `taskCountMax`.|Yes||
|`taskCountMin`|Minimum number of Kinesis ingestion tasks. When you enable the auto scaler, Druid ignores the value of `taskCount` in `IOConfig` and uses `taskCountMin` for the initial number of tasks to launch.|Yes||
|`minTriggerScaleActionFrequencyMillis`|Minimum time interval between two scale actions.| No|600000|
|`autoScalerStrategy`|The algorithm of `autoScaler`. Druid only supports the `lagBased` strategy. See [Lag based autoscaler strategy related properties](#lag-based-autoscaler-strategy-related-properties) for more information.|No|Defaults to `lagBased`.|

### Lag based autoscaler strategy related properties

> Unlike the Kafka
> indexing service, Kinesis reports lag metrics measured in time difference in milliseconds between the current sequence number and latest sequence number, rather than message count.

Reviewer: ```suggestion
Unlike the Kafka indexing service, Kinesis reports lag metrics measured in time difference in milliseconds between the current sequence number and latest sequence number, rather than message count.
```

  ~ under the License.
  -->

When you enable the Kinesis indexing service, you can configure supervisors on the Overlord to manage the creation and lifetime of Kinesis indexing tasks. These indexing tasks read events using Kinesis' own shard and sequence number mechanism to guarantee exactly-once ingestion. The supervisor oversees the state of the indexing tasks to coordinate handoffs, manage failures, and ensure that scalability and replication requirements are maintained.

To use the Kinesis indexing service, load the `druid-kinesis-indexing-service` core Apache Druid extension. See [Loading extensions](../../configuration/extensions.md#loading-extensions) for more information.

> Before you deploy the Kinesis extension to production, read the [Kinesis known issues](#kinesis-known-issues).

> ## Submitting a supervisor spec

Author: 100% agree! I will save it for a different PR.

> To use the Kinesis indexing service, load the `druid-kinesis-indexing-service` extension on both the Overlord and the Middle Managers. Druid starts a supervisor for a datasource when you submit a supervisor spec. Submit your supervisor spec to the following endpoint:

Reviewer: nit: link to supervisors API doc?

> curl -X POST -H 'Content-Type: application/json' -d @supervisor-spec.json http://localhost:8090/druid/indexer/v1/supervisor

Reviewer: imo it would be better to use a minimal inline supervisor spec in the payload rather than to link to a file. Using tabbed cURL & HTTP examples. See the sample request here: https://github.com/apache/druid/pull/14492/

### `tuningConfig`

> The `tuningConfig` parameter is optional. If you don't specify `tuningConfig`, Druid uses default parameters.

Reviewer: `tuningConfig` is an object (JSON object), not a parameter. If you don't specify... Druid uses the defaults or the default configurations.

Author: Updated

#### IndexSpec

|**Field**|**Type**|**Description**|**Required**|**Default**|
|---------|--------|---------------|------------|-----------|
|`bitmap`|Object|Compression format for bitmap indexes. Should be a JSON object. See [Bitmap types](#bitmap-types) for options.|No|Roaring|
|`dimensionCompression`|String|Compression format for dimension columns. Choose from `LZ4`, `LZF`, or `uncompressed`.|No|`LZ4`|
|`metricCompression`|String|Compression format for primitive type metric columns. Choose from `LZ4`, `LZF`, `uncompressed`, or `none`.|No|`LZ4`|
|`longEncoding`|String|Encoding format for metric and dimension columns with type long. Choose from `auto` or `longs`. `auto` encodes the values using sequence number or lookup table depending on column cardinality and stores them with variable sizes. `longs` stores the value as is with 8 bytes each.|No|`longs`|

##### Bitmap types

|**Compression scheme**|**Field**|**Type**|**Description**|**Required**|
|----------------------|---------|--------|---------------|------------|

> |Roaring|`type`|String|Must be `roaring`.|Yes|

Reviewer: The descriptions here are not doing a lot of work. We should either link out to help folks understand how to choose a compression scheme or add some info so people know what to choose. Also is one or the other of these the default. (This might be for another PR)

|`CONNECTING_TO_STREAM` (first iteration only)|`RUNNING`|The supervisor is trying to connect to the stream and update partition data.|
|`DISCOVERING_INITIAL_TASKS` (first iteration only)|`RUNNING`|The supervisor is discovering already-running tasks.|
|`CREATING_TASKS` (first iteration only)|`RUNNING`|The supervisor is creating tasks and discovering state.|
|`RUNNING`|`RUNNING`|The supervisor has started tasks and is waiting for taskDuration to elapse.|
|`SUSPENDED`|`SUSPENDED`|The supervisor has been suspended.|
|`STOPPING`|`STOPPING`|The supervisor is stopping.|

On each iteration of the supervisor's run loop, the supervisor completes the following tasks in sequence:

1. Fetch the list of shards from Kinesis and determine the starting sequence number for each shard (either based on the
  last processed sequence number if continuing, or starting from the beginning or ending of the stream if this is a new stream).
2. Discover any running indexing tasks that are writing to the supervisor's datasource and adopt them if they match
  the supervisor's configuration, else signal them to stop.

> 3. Send a status request to each supervised task to update our view of the state of the tasks under our supervision.

Reviewer 2: ```suggestion
3. Send a status request to each supervised task to update the view of the state of the tasks under supervision.
```

that is, once it has completed a full execution without encountering any issues - `detailedState` will show a `RUNNING`
state until it is stopped, suspended, or hits a failure threshold and transitions to an unhealthy state.

### Updating existing supervisors

You can send a `POST` request to `/druid/indexer/v1/supervisor` to update the existing supervisor spec.
Calling this endpoint when there is already an existing supervisor for the same datasource will cause:

- The running supervisor to signal its managed tasks to stop reading and begin publishing.
- The running supervisor to exit.
- A new supervisor to be created using the configuration provided in the request body. This supervisor retains the
existing publishing tasks and creates new tasks starting at the sequence numbers the publishing tasks ended on.

Seamless schema migrations can thus be achieved by simply submitting the new schema using this endpoint.

### Suspending and resuming supervisors

To suspend a supervisor, send a `POST` request to `/druid/indexer/v1/supervisor/<supervisorId>/suspend`.
To resume a supervisor, send a `POST` request to `/druid/indexer/v1/supervisor/<supervisorId>/resume`.

> The supervisor itself will still be operating and emitting logs and metrics,

Reviewer 2: ```suggestion
The supervisor still operates and emits logs and metrics,
```

> it will just ensure that no indexing tasks are running until the supervisor is resumed.

Reviewer 2: ```suggestion
and it ensures that no indexing tasks are running until the supervisor resumes.
```

Note that the supervisor must be running for this endpoint to be available.

### Terminating supervisors

Sending a `POST` request to `/druid/indexer/v1/supervisor/<supervisorId>/terminate` terminates a supervisor and causes
all associated indexing tasks managed by this supervisor to immediately stop and begin
publishing their segments. This supervisor will still exist in the metadata store and its history may be retrieved
with the supervisor history API, but will not be listed in the 'get supervisors' API response nor can its configuration
or status report be retrieved. The only way this supervisor can start again is by submitting a functioning supervisor
spec to the create API.

### Capacity planning

> Kinesis indexing tasks run on MiddleManagers and are limited by the resources available in the MiddleManager

Reviewer 2: ```suggestion
Kinesis indexing tasks run on Middle Managers and are limited by the resources available in the Middle Manager
```

cluster. In particular, you should make sure that you have sufficient worker capacity (configured using the
`druid.worker.capacity` property) to handle the configuration in the supervisor spec. Note that worker capacity is
shared across all types of indexing tasks, so you should plan your worker capacity to handle your total indexing load, such as batch processing, streaming tasks, and merging tasks. If your workers run out of capacity, Kinesis indexing tasks queue and wait for the next available worker. This may cause queries to return partial results but will not result in data loss (assuming the tasks run before Kinesis purges those sequence numbers).

A running task can be in one of two states: reading or publishing. A task remains in reading state for

> `taskDuration`, at which point it transitions to publishing state. A task remains in publishing state for as long

Reviewer 2: ```suggestion
the period defined in `taskDuration`, at which point it transitions to publishing state. A task remains in publishing state for as long
```

as it takes to generate segments, push segments to deep storage, and have them loaded and served by a Historical process
(or until `completionTimeout` elapses).

The number of reading tasks is controlled by `replicas` and `taskCount`. In general, there will be `replicas * taskCount`

> reading tasks, the exception being if `taskCount > {numKinesisShards}` in which case `{numKinesisShards}` tasks will

Reviewer 2: ```suggestion
reading tasks. An exception occurs if `taskCount > {numKinesisShards}` in which case `{numKinesisShards}` tasks are
```

> be used instead. When `taskDuration` elapses, these tasks will transition to publishing state and `replicas * taskCount`

Reviewer 2: ```suggestion
used instead. When `taskDuration` elapses, these tasks transition to publishing state and `replicas * taskCount`
```

> new reading tasks will be created. To allow for reading tasks and publishing tasks to run concurrently, there

Reviewer 2: ```suggestion
new reading tasks are created. To allow for reading tasks and publishing tasks to run concurrently, there
```

This value is for the ideal situation in which there is at most one set of tasks publishing while another set is reading.
In some circumstances, it is possible to have multiple sets of tasks publishing simultaneously. This would happen if the
time-to-publish (generate segment, push to deep storage, loaded on Historical) is greater than `taskDuration`. This is a valid

> scenario (correctness-wise) but requires additional worker capacity to support. In general, it is a good idea to have

Reviewer 2: ```suggestion
and correct scenario but requires additional worker capacity to support. In general, it is a good idea to have
```

`taskDuration` be large enough that the previous set of tasks finishes publishing before the current set begins.

### Supervisor persistence

When a supervisor spec is submitted via the `POST /druid/indexer/v1/supervisor` endpoint, it is persisted in the
configured metadata database. There can only be a single supervisor per datasource, and submitting a second spec for
the same datasource overwrites the previous one.

When an Overlord gains leadership, either by being started or as a result of another Overlord failing, it spawns
a supervisor for each supervisor spec in the metadata database. The supervisor then discovers running Kinesis indexing
tasks and attempts to adopt them if they are compatible with the supervisor's configuration. If they are not
compatible because they have a different ingestion spec or shard allocation, the tasks are killed and the
supervisor creates a new set of tasks. In this way, the supervisors are persistent across Overlord restarts and

> fail-overs.

Reviewer 2: ```suggestion
failovers.
```

tombstone marker in the database (to prevent the supervisor from being reloaded on a restart) and then gracefully
shuts down the currently running supervisor. When you shut a supervisor down in this way, it instructs its
managed tasks to stop reading. The tasks begin publishing their segments immediately. The call to the shutdown
endpoint returns after all tasks have been signalled to stop but before the tasks finish publishing their segments.

### Schema and configuration changes

Schema and configuration changes are handled by submitting the new supervisor spec via the same
`POST /druid/indexer/v1/supervisor` endpoint used to initially create the supervisor. The Overlord initiates a
graceful shutdown of the existing supervisor which causes the tasks being managed by that supervisor to stop reading
and begin publishing their segments. A new supervisor is then started which creates a new set of tasks that start reading from the sequence numbers where the previous now-publishing tasks left off, but using the updated schema.
In this way, configuration changes can be applied without requiring any pause in ingestion.

### Deployment notes

> Each Kinesis Indexing Task puts events consumed from Kinesis shards assigned to it in a single segment for each segment

Reviewer 2: ```suggestion
Each Kinesis indexing task puts events consumed from Kinesis shards assigned to it in a single segment for each segment
```

## Resharding

When changing the shard count for a Kinesis stream, there is a window of time around the resharding operation with early shutdown of Kinesis ingestion tasks and possible task failures.

The early shutdowns and task failures are expected. They occur because the supervisor updates the shard to task group mappings as shards are closed and fully read. This ensures that tasks are not running
with an assignment of closed shards that have been fully read and balances distribution of active shards across tasks.

This window with early task shutdowns and possible task failures concludes when:

- All closed shards have been fully read and the Kinesis ingestion tasks have published the data from those shards, committing the "closed" state to metadata storage.

> - Any remaining tasks that had inactive shards in the assignment have been shutdown. These tasks would have been created before the closed shards were completely drained.

Reviewer 2: ```suggestion
- Any remaining tasks that had inactive shards in the assignment have been shut down. These tasks would have been created before the closed shards were completely drained.
```
